File: GW3/LogFileHandling/Query/Tokens/TokenAnd.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenAnd : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();

            return ((parser.PeekChar() == '&' && parser.PeekChar(1) == '&') || parser.PeekString().ToLower() == "and");
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            if (parser.PeekString().ToLower() == "and")
            {
                parser.NextString();
                return new TokenAnd { Value = "&&" };
            }
            return new TokenAnd { Value = "" + parser.NextChar() + parser.NextChar() };
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenAnd.ts
class TokenAnd extends Token
{

    public tokenType: TokenType = TokenType.TokenAnd;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[
        { suggestion: "and" },
        { suggestion: "&&" }
    ];

    public canBeUsed(parser: QueryParser): boolean
    {
        parser.skipSpaces();
        return ((parser.peekChar() == '&' && parser.peekChar(1) == '&') || parser.peekString().toLowerCase() == "and");
    }

    public extract(parser: QueryParser): Token
    {
        parser.skipSpaces();
        var token = new TokenAnd;
        if (parser.peekString().toLowerCase() == "and")
        {
            parser.nextString();
            token.value = "&&";
        }
        else
            token.value = parser.nextChar() + parser.nextChar();
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[]
    {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined')
        {
            switch (nextToken.tokenType)
            {
                case TokenType.TokenName:
                    return [new TokenCompare()];
                case TokenType.TokenOpenParenthesis:
                    return [new TokenName(), new TokenOpenParenthesis()];
            }
        }
        return [];
    }
}